compiler/rustc_borrowck/src/diagnostics/explain_borrow.rs

```diff
@@ -359,19 +356,37 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
         let borrow_region_vid = borrow.region;
         debug!(?borrow_region_vid);
 
-        let region_sub = self.regioncx.find_sub_region_live_at(borrow_region_vid, location);
+        let mut region_sub = self.regioncx.find_sub_region_live_at(borrow_region_vid, location);
         debug!(?region_sub);
 
-        match find_use::find(body, regioncx, tcx, region_sub, location) {
+        let mut use_location = location;
+        let mut use_in_later_iteration_of_loop = false;
+
+        if region_sub == borrow_region_vid {
+            // When `region_sub` is the same as `borrow_region_vid` (the location where the borrow is
+            // issued is the same location that invalidates the reference), this is likely a loop iteration
+            // - in this case, try using the loop terminator location in `find_sub_region_live_at`.
```

Comment:
The code suggests that we can only have a loop iteration if `region_sub == borrow_region_vid`.  Could you explain this assumption?

Reply:
The assumption might be explained intuitively using the hypothetical logic of the following problem:
Given point `A`, find point `B` such that there is a path from point `A` to point `B`. If `B` equals `A`, the path is a cycle.

Likewise, consider the line in question:
```
let mut region_sub = self.regioncx.find_sub_region_live_at(borrow_region_vid, location);
```
based on the implementation of the function `find_sub_region_live_at`:
```
/// Finds some region R such that `fr1: R` and `R` is live at `elem`.
find_sub_region_live_at(&self, fr1: RegionVid, elem: Location) -> RegionVid
    self.find_constraint_paths_between_regions(fr1, |r| {
        // First look for some `r` such that `fr1: r` and `r` is live at `elem`
        self.liveness_constraints.contains(r, elem)
    })
    .map(|(_path, r)| r)
    .unwrap()
}
```
The function can be rephrased to the following problem:
Given region fr1 (`A`), find region R to which 'elem' belongs (`B`) so that there is a constraint path from region fr1 (`A`) to region R (`B`).  
If R (`B`) equals fr1 (`A`), then the constraint path is a loop.

Therefor, if region_sub (corresponding to `B`) equals to borrow_region_vid (corresponding to `A`), it is in a loop.

Reply:
Thanks for the explanation.